ccl/backupccl: TestBackupRestoreTenant failed

ccl/backupccl.TestBackupRestoreTenant [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/11081075?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsGcpLinuxX8664BigVm_CclUnitTests/11081075?buildTab=artifacts#/) on master @ [57b6b702a0403fca0d8edce22d8758f6847d6913](https://github.com/cockroachdb/cockroach/commits/57b6b702a0403fca0d8edce22d8758f6847d6913):

Fatal error:

```
panic: test timed out after 14m55s

```
Stack: 

```
goroutine 685140 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2036 +0x8e
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x32
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestBackupRestoreTenant
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/ea72cae9b60f53a3f66a94ca0ac417fd/logTestBackupRestoreTenant1877002037
    test_log_scope.go:81: use -show-logs to present logs inline
=== RUN   TestBackupRestoreTenant/invalid
=== RUN   TestBackupRestoreTenant/non-existent
=== RUN   TestBackupRestoreTenant/restore-all-from-cluster-backup
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/disaster-recovery
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestBackupRestoreTenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-30158

--- a/pkg/ccl/backupccl/backup_test.go
+++ b/pkg/ccl/backupccl/backup_test.go
@@ -6943,7 +6943,6 @@ func TestBackupRestoreInsideMultiPodTenant(t *testing.T) {
 func TestBackupRestoreTenant(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	skip.WithIssue(t, 107669)
 
 	params := base.TestClusterArgs{ServerArgs: base.TestServerArgs{
 		Knobs: base.TestingKnobs{
@@ -7353,7 +7352,7 @@ func TestBackupRestoreTenant(t *testing.T) {
 		restoreDB.Exec(t, `ALTER TENANT [20] STOP SERVICE`)
 		restoreDB.Exec(t, `DROP TENANT [20] IMMEDIATE`)
 
-		restoreDB.Exec(t, `RESTORE TENANT 11 FROM 'nodelocal://1/clusterwide' WITH virtual_cluster_name = 'tenant-20'`)
+		restoreDB.Exec(t, `RESTORE TENANT 11 FROM 'nodelocal://1/clusterwide' WITH virtual_cluster = '20', virtual_cluster_name = 'tenant-20'`)
 
 		tenantID = roachpb.MustMakeTenantID(20)
 		if err := restoreTC.Server(0).(*server.TestServer).WaitForTenantReadiness(ctx, tenantID); err != nil {

--- a/pkg/server/testserver.go
+++ b/pkg/server/testserver.go
@@ -72,6 +72,7 @@ import (
 	addrutil "github.com/cockroachdb/cockroach/pkg/util/netutil/addr"
 	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
 	"github.com/cockroachdb/errors"
 	"github.com/gogo/protobuf/proto"
@@ -1083,20 +1084,33 @@ func (t *TestTenant) HTTPAuthServer() interface{} {
 	return t.t.authentication
 }
 
-// WaitForTenantReadiness is part of TestServerInterface.
-func (ts *TestServer) WaitForTenantReadiness(ctx context.Context, tenantID roachpb.TenantID) error {
-	// Restarting the watcher forces a new initial scan which is
-	// faster than waiting out the closed timestamp interval
-	// required to see new updates.
-	ts.node.tenantInfoWatcher.TestingRestart()
-
-	log.Infof(ctx, "waiting for rangefeed to catch up with record for tenant %v", tenantID)
-
+func (ts *TestServer) waitForTenantReadinessImpl(
+	ctx context.Context, tenantID roachpb.TenantID,
+) error {
 	_, infoWatcher, err := ts.node.waitForTenantWatcherReadiness(ctx)
 	if err != nil {
 		return err
 	}
 
+	// Ignore the tenant capabilities entry as it may be served from a stale
+	// version of the in-RAM cache. We use the returned channel to monitor when
+	// the watcher has completed its initial scan and hydrated the in-RAM cache
+	// with the most up-to-date values.
+	_, infoCh, _ := infoWatcher.GetInfo(tenantID)
+
+	// Restarting the watcher forces a new initial scan which is faster than
+	// waiting out the closed timestamp interval required to see new updates.
+	ts.node.tenantInfoWatcher.TestingRestart()
+
+	log.Infof(ctx, "waiting for rangefeed to catch up with record for tenant %v", tenantID)
+
+	// Wait for the watcher to handle the complete update from the initial scan
+	// and notify our previously registered listener.
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-infoCh:
+	}
 	for {
 		info, infoCh, found := infoWatcher.GetInfo(tenantID)
 		if found && info.ServiceMode != mtinfopb.ServiceModeNone {
@@ -1111,6 +1125,15 @@ func (ts *TestServer) WaitForTenantReadiness(ctx context.Context, tenantID roach
 			continue
 		}
 	}
+}
+
+// WaitForTenantReadiness is part of serverutils.TenantControlInterface..
+func (ts *TestServer) WaitForTenantReadiness(ctx context.Context, tenantID roachpb.TenantID) error {
+	// Two minutes should be sufficient for the in-RAM caches to be hydrated with
+	// the tenant record.
+	return timeutil.RunWithTimeout(ctx, "waitForTenantReadiness", 2*time.Minute, func(ctx context.Context) error {
+		return ts.waitForTenantReadinessImpl(ctx, tenantID)
+	})
 }
 
 // StartTenant is part of the serverutils.TenantControlInterface.

